packages/common/src/concurrency/ReadWriteLock.ts

```diff
@@ -0,0 +1,87 @@
+import { EventEmitter } from 'events';
+
+import { Mutex } from './Mutex';
+
+export class ReadWriteLock {
+  #mutex: Mutex;
+  #emitter: EventEmitter;
+  #readCount: number;
+  #writeCount: number;
+  #waitingWriters: number;
+
+  constructor() {
+    this.#mutex = new Mutex();
+    this.#emitter = new EventEmitter();
+    this.#readCount = 0;
+    this.#writeCount = 0;
+    this.#waitingWriters = 0;
+  }
+
+  async acquireReadLock() {
+    await this.#mutex.lock();
+    try {
+      if (this.#writeCount > 0 || this.#waitingWriters > 0) {
+        const readPromise = new Promise<void>((resolve) => {
+          const handler = () => {
+            this.#emitter.removeListener('writeComplete', handler);
+            resolve();
+          };
+          this.#emitter.on('writeComplete', handler);
```

Comment:
I think you could do this:
```typescript
this.#emitter.once('writeComplete', resolve)
```
So you could get rid of defining a handler and dealing with the removal
[reference](https://nodejs.org/api/events.html#emitteronceeventname-listener)